frontend/src/metabase/actions/components/ActionForm/FormFieldEditor.tsx

```diff
@@ -42,11 +43,24 @@ function FormFieldEditor({
   const inputTypes = useMemo(getInputTypes, []);
 
   const handleChangeFieldType = (nextFieldType: FieldType) => {
-    const [defaultInputType] = inputTypes[nextFieldType];
+    // When field type changes, we automatically set the first available input type
+    const [nextInputType] = inputTypes[nextFieldType];
+
+    const shouldResetOptions =
+      nextFieldType !== fieldSettings.fieldType ||
+      !inputTypeHasOptions(nextInputType.value);
```

Comment:
I think auto-erasing user-defined options here will be frustrating to users. What if, instead of clearing the options any time the fieldType changes, if we filter them based on validity?

**Scenario 1**
- I define a list of options: 1,2,3,4,5
- oh crap, i forgot to change the field type to number
- change fieldType to number
- all my options are gone 😢 

**Scenario 2**
- I select number fieldType
- I define a list of options: 1,2,3,4,5
- I realize that my column is actually a text column
- I change the fieldType to string
- all my options are gone 😢 

**What if instead we filter valid options**
- I define text options: 1,2,3,4,5, "none of the above"
- I change the fieldType to number
- now my options are: 1,2,3,4,5
- makes sense 😁 

Also, if you change from number -> string, everything remains valid.

Reply:
That's an excellent idea, thank you! I think I've managed to achieve this. Please take a look at this demo and LMK if it matches your idea:

https://user-images.githubusercontent.com/17258145/223750916-6d74f5bc-8bea-4991-93c1-7602cd0fb36f.mp4